### PR TITLE
Add plugin adapters.

### PR DESCRIPTION
**Pull request by Cybermaxke**
[API](https://github.com/SpongePowered/SpongeAPI/pull/1990) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/2238) | **Forge** | Vanilla | [KotlinAdapter](https://github.com/Cybermaxke/KotlinAdapter)

### CODE REVIEW

File: src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java [hunk 1]
```diff
@@ -273,15 +280,44 @@ public void constructMod(FMLConstructionEvent event) {
...
+
+            Injector injector = spongeInjector.getParent().createChildInjector(new PluginModule((PluginContainer) this));
+
+            final ImmutableList.Builder<Module> modules = ImmutableList.builder();
+            // Check for plugin specified modules
+            final List<Type> injectionModuleTypes = (List<Type>) this.descriptor.get("injectionModules");
+            if (injectionModuleTypes != null && injectionModuleTypes.size() > 0) {
+                for (Type injectionModuleType : injectionModuleTypes) {
+                    final Class<?> moduleClass = Class.forName(injectionModuleType.getClassName(), true, modClassLoader);
+                    modules.add((Module) moduleClass.newInstance());
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java]: Create it using injector, so modules can depend on sponge injectable things

File: src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java [hunk 2]
```diff
@@ -273,15 +280,44 @@ public void constructMod(FMLConstructionEvent event) {
...
+                }
+            }
+
+            final PluginAdapter adapter = adapterClass == PluginAdapter.Default.class ? DefaultPluginAdapter.INSTANCE :
+                    (PluginAdapter) adapterClass.newInstance();
```
Comment (XakepSDK) [inline src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java]: Inject? But i'm not sure about here

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/mod/plugin/SpongeModPluginContainer.java]: I will create the instance with the parent injector, adapters should be reusable for multiple plugins. Although it's currently not implemented this way.

**Comment by ImMorpheus** [comment]
https://github.com/SpongePowered/SpongeAPI/pull/1990#issuecomment-650779079

**Pull request closed without merging**